File: apps/portal/src/app/features/model-gateway/credential-reference.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

const ENV_REFERENCE = /^env:NOVA_PROVIDER_[A-Z0-9_]+$/;
const VAULT_REFERENCE =
  /^vault:provider-secret:[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;
const BEARER = /bearer\s+/i;
const PRIVATE_KEY = /-----BEGIN (?:RSA )?PRIVATE KEY-----/i;
const JSON_SECRET = /^\s*\{.*"(?:api[_-]?key|secret|token)".*\}\s*$/is;
const PLAINTEXT_KEY = /(?:sk-[a-z0-9]{20,}|api[_-]?key\s*[:=])/i;

export function looksLikePlaintextSecret(value: string): boolean {
  return (
    BEARER.test(value) ||
    PRIVATE_KEY.test(value) ||
    JSON_SECRET.test(value) ||
    PLAINTEXT_KEY.test(value)
  );
}

export function isValidCredentialReference(value: string | null | undefined): boolean {
  if (value == null) {
    return true;
  }
  const trimmed = value.trim();
  if (!trimmed) {
    return true;
  }
  if (looksLikePlaintextSecret(trimmed)) {
    return false;
  }
  return ENV_REFERENCE.test(trimmed) || VAULT_REFERENCE.test(trimmed);
}

export function credentialReferenceValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const value = typeof control.value === 'string' ? control.value : '';
    return isValidCredentialReference(value) ? null : { credentialReference: true };
  };
}

File: services/platform-api/src/main/java/ai/nova/platform/modelgateway/validation/CredentialReferenceValidator.java
package ai.nova.platform.modelgateway.validation;

import java.util.regex.Pattern;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;

import ai.nova.platform.modelgateway.entity.AiProviderType;
import ai.nova.platform.web.error.ApiException;

@Component
public class CredentialReferenceValidator {

    private static final Pattern ENV_REFERENCE =
            Pattern.compile("^env:NOVA_PROVIDER_[A-Z0-9_]+$");
    private static final Pattern BEARER = Pattern.compile("(?i)bearer\\s+");
    private static final Pattern PRIVATE_KEY = Pattern.compile("(?i)-----BEGIN (?:RSA )?PRIVATE KEY-----");
    private static final Pattern JSON_SECRET = Pattern.compile("^\\s*\\{.*\"(?:api[_-]?key|secret|token)\".*\\}\\s*$", Pattern.DOTALL);
    private static final Pattern PLAINTEXT_KEY = Pattern.compile("(?i)(?:sk-[a-z0-9]{20,}|api[_-]?key\\s*[:=])");

    public void validate(String credentialReference, AiProviderType providerType) {
        if (providerType == AiProviderType.DETERMINISTIC_LOCAL) {
            if (credentialReference != null && !credentialReference.isBlank()) {
                throw new ApiException(
                        HttpStatus.BAD_REQUEST,
                        "INVALID_CREDENTIAL_REFERENCE",
                        "Deterministic local providers must not have credentials");
            }
            return;
        }
        if (credentialReference == null || credentialReference.isBlank()) {
            return;
        }
        String trimmed = credentialReference.trim();
        rejectSuspicious(trimmed);
        if (!ENV_REFERENCE.matcher(trimmed).matches()) {
            throw new ApiException(
                    HttpStatus.BAD_REQUEST,
                    "INVALID_CREDENTIAL_REFERENCE",
                    "Credential reference must match env:NOVA_PROVIDER_<NAME>");
        }
    }

    public void rejectSuspicious(String value) {
        if (value == null || value.isBlank()) {
            return;
        }
        if (BEARER.matcher(value).find()
                || PRIVATE_KEY.matcher(value).find()
                || JSON_SECRET.matcher(value).find()
                || PLAINTEXT_KEY.matcher(value).find()) {
            throw new ApiException(
                    HttpStatus.BAD_REQUEST,
                    "INVALID_CREDENTIAL_REFERENCE",
                    "Plaintext secrets are not allowed");
        }
    }
}
